File: src/creational/abstract-factory/abstract_factory.py

# Abstract Factory interface
class AbstractFactory:
    def create_product_a(self):
        pass

    def create_product_b(self):
        pass

# Concrete Factory 1
class ConcreteFactory1(AbstractFactory):
    def create_product_a(self):
        return ConcreteProductA1()

    def create_product_b(self):
        return ConcreteProductB1()

# Concrete Factory 2
class ConcreteFactory2(AbstractFactory):
    def create_product_a(self):
        return ConcreteProductA2()

    def create_product_b(self):
        return ConcreteProductB2()

# Abstract Product A
class AbstractProductA:
    def use_product_a(self):
        pass

# Concrete Product A1
class ConcreteProductA1(AbstractProductA):
    def use_product_a(self):
        print("Using ConcreteProductA1")

# Concrete Product A2
class ConcreteProductA2(AbstractProductA):
    def use_product_a(self):
        print("Using ConcreteProductA2")

# Abstract Product B
class AbstractProductB:
    def use_product_b(self):
        pass

# Concrete Product B1
class ConcreteProductB1(AbstractProductB):
    def use_product_b(self):
        print("Using ConcreteProductB1")

# Concrete Product B2
class ConcreteProductB2(AbstractProductB):
    def use_product_b(self):
        print("Using ConcreteProductB2")

# Client code
class Client:
    def __init__(self, factory):
        self._product_a = factory.create_product_a()
        self._product_b = factory.create_product_b()

    def use_products(self):
        self._product_a.use_product_a()
        self._product_b.use_product_b()

if __name__ == "__main__":
    print("Client: Testing with ConcreteFactory1")
    client1 = Client(ConcreteFactory1())
    client1.use_products()

    print("Client: Testing with ConcreteFactory2")
    client2 = Client(ConcreteFactory2())
    client2.use_products()

File: src/creational/abstract-factory/AbstractFactory.cs

using System;

// Abstract Factory interface
public interface IAbstractFactory
{
    IAbstractProductA CreateProductA();
    IAbstractProductB CreateProductB();
}

// Concrete Factory 1
public class ConcreteFactory1 : IAbstractFactory
{
    public IAbstractProductA CreateProductA()
    {
        return new ConcreteProductA1();
    }

    public IAbstractProductB CreateProductB()
    {
        return new ConcreteProductB1();
    }
}

// Concrete Factory 2
public class ConcreteFactory2 : IAbstractFactory
{
    public IAbstractProductA CreateProductA()
    {
        return new ConcreteProductA2();
    }

    public IAbstractProductB CreateProductB()
    {
        return new ConcreteProductB2();
    }
}

// Abstract Product A
public interface IAbstractProductA
{
    void UseProductA();
}

// Concrete Product A1
public class ConcreteProductA1 : IAbstractProductA
{
    public void UseProductA()
    {
        Console.WriteLine("Using ConcreteProductA1");
    }
}

// Concrete Product A2
public class ConcreteProductA2 : IAbstractProductA
{
    public void UseProductA()
    {
        Console.WriteLine("Using ConcreteProductA2");
    }
}

// Abstract Product B
public interface IAbstractProductB
{
    void UseProductB();
}

// Concrete Product B1
public class ConcreteProductB1 : IAbstractProductB
{
    public void UseProductB()
    {
        Console.WriteLine("Using ConcreteProductB1");
    }
}

// Concrete Product B2
public class ConcreteProductB2 : IAbstractProductB
{
    public void UseProductB()
    {
        Console.WriteLine("Using ConcreteProductB2");
    }
}

// Client code
class Client
{
    private IAbstractProductA _productA;
    private IAbstractProductB _productB;

    public Client(IAbstractFactory factory)
    {
        _productA = factory.CreateProductA();
        _productB = factory.CreateProductB();
    }

    public void UseProducts()
    {
        _productA.UseProductA();
        _productB.UseProductB();
    }
}

class Program
{
    static void Main()
    {
        Console.WriteLine("Client: Testing with ConcreteFactory1");
        Client client1 = new Client(new ConcreteFactory1());
        client1.UseProducts();

        Console.WriteLine("Client: Testing with ConcreteFactory2");
        Client client2 = new Client(new ConcreteFactory2());
        client2.UseProducts();
    }
}
